app/controllers/api/v0/api_controller.rb

```diff
@@ -29,6 +29,47 @@ def auth_results
     render json: { status: "ok" }
   end
 
+  def user_qualification_data
+    date = cutoff_date
+    return render json: { error: 'Invalid date format. Please provide an iso8601 date string.' }, status: :bad_request unless date.present?
+    return render json: { error: 'You cannot request qualification data for a future date.' }, status: :bad_request if date > Date.current
+
+    user = User.find(params.require(:user_id))
+    return render json: [] unless user.person.present?
+
+    # Compile singles
+    best_singles_by_cutoff = user.person.best_singles_by(date)
+    single_qualifications = best_singles_by_cutoff.map do |event, time|
+      qualification_data(event, :single, time, date)
+    end
+
+    # Compile averages
+    best_averages_by_cutoff = user.person&.best_averages_by(date)
+    average_qualifications = best_averages_by_cutoff.map do |event, time|
+      qualification_data(event, :average, time, date)
+    end
+
+    render json: single_qualifications + average_qualifications
+  end
+
+  private def cutoff_date
+    if params[:date].present?
+      Date.safe_parse(params[:date])
+    else
+      Date.current
+    end
+  end
+
+  private def qualification_data(event, type, time, date)
+    raise ArgumentError.new("'type' may only contain the symbols `:single` or `:average`") unless [:single, :average].include?(type)
```

Comment:
Not sure if you need this in a `private` method definition. This kind of error message is helpful in external, world-facing APIs but since this is by definition code that only we call ourselves, it feels a bit overkill.

You don't have to delete this line to get my approval, just noting it here.

Reply:
My thinking here is to rather fail loudly when someone gives an invalid input, than return `nil` or similar and then someone has to dig to figure out where the `nil` is coming from. Are there alternate approaches you'd suggest?

Reply:
Use a strongly typed programming language.
I appreciate your thinking, but the ultimate consequence would be to bloat our entire codebase by at least factor 1.5 by introducing is_instance? checks everywhere.

Reply:
> Use a strongly typed programming language.
TypeRuby :pray: 

> I appreciate your thinking, but the ultimate consequence would be to bloat our entire codebase by at least factor 1.5 by introducing is_instance? checks everywhere.

Not opposed to forgoing these types of checks, my arguments in favour of them would be: 
- I only see this being relevant/necessary in cases where we are defining some custom input values/symbols, which should be pretty rare.
- It's only a single line of code at the start of a function where it's needed, so the impact should hopefully be less than 1.5x our codebase - and it seems like the time saved on developers chasing obscure errors may be beneficial
- I assume that even in a strongly typed language you have to have some boilerplate somewhere that says "hey here are the explicitly-accepted input values for this custom argument"